### DISCUSSION TITLE
Lazy dependencies and Environments

### DISCUSSION THREAD

**Discussion by bielikb** [pointfreeco/swift-composable-architecture#412] [Q&A] [answered]
Hi,

Ive got (probably) amateur question as I'm exploring the Composable Architecture.
In the following graph, you can see the flow/view hierarchy of the Scrumdinger app that I tried to migrate to Composable Architecture.

![graph](https://user-images.githubusercontent.com/4924709/110112861-91e09f80-7db2-11eb-8594-03996b8145d1.png)

Most of the views in this graph have their own environments with their own env dependencies specified, yet not all of the environments should be wandering around  throughout the whole life-cycle of the app.

Currently that's the case in the project Im working on. All environments that are used within the app are initialized at the top level. 

_**Is there a way how to actually defer the initialization time of specific environment down in the view hierarchy?**_

Let's take an exact example:
I've defined following environment for `MeetingView` along with `meetingReducer`.

```swift
struct MeetingEnvironment {
    let scrumTimer: ScrumTimer
    ....
}
```

In the parent reducer for `DetailView` I'm using pullback API to bind the state and action to the parent reducer.

```swift
let detailViewReducer = Reducer<DetailViewState, DetailViewAction, DetailViewEnvironment>.combine(
    meetingViewReducer.pullback(state: \DetailViewState.meetingState,
                                action: /DetailViewAction.meetingAction,
                                environment: { $0.meetingViewEnvironment }),
    ...
)
```

My issue here is that If I wouldn't keep the reference to the instance of `MeetingEnvironment ` that initialises the `ScrumTimer`, the `ScrumTimer` would be deinitialised as soon as the `pullback` closure would be called

Maybe Im doing something wrong, so feel free to correct me. Thanks in advance!

Best,
Boris

Source code:
[detailViewReducer](https://github.com/bielikb/scrumdinger/blob/aff2606ec16bcc29105a3230d98815334213efde/Scrumdinger/DetailView.swift#L40)
[MeetingViewEnvironment](https://github.com/bielikb/scrumdinger/blob/aff2606ec16bcc29105a3230d98815334213efde/Scrumdinger/MeetingView.swift#L38)

**Comment by stephencelis** [comment, marked as answer]
> My issue here is that If I wouldn't keep the reference to the instance of `MeetingEnvironment ` that initialises the `ScrumTimer`, the `ScrumTimer` would be deinitialised as soon as the `pullback` closure would be called

Do you have an example of this? `Reducer.pullback` captures a _transformation_ of a global environment into a more local one, but that global environment should be as long living as your app's store, so the `scrumTimer` shouldn't ever be deinitialized.

**Reply by bielikb** [reply]
Actually, that's the case in the current stage of my implementation. All the environments are initialised as soon as the [appEnvironment](https://github.com/bielikb/scrumdinger/blob/23c35919453186cdfe1e9d2dd074654cb4fd3a1e/Scrumdinger/ScrumdingerApp.swift#L66) is instantiated. 
The "env dependency hierarchy" copies that of the view hierarchy.

I suppose my question here is, whether there's a way to instantiate & deinitialise the environments just as they're needed, and to avoid keeping them around throughout the whole app life-cycle.

I could mark the `ScrumTimer` property as `lazy`, yet I wondered if there's any other strategy in TCA that would allow me to "pullback & keep" the newly created instance of environment just for the time it's needed.
One of the possible ways of doing so would be to introduce optionality, but that would overcomplicate the whole API.

I guess, currently Im just probably overthinking the initialisation strategy for sub-environments held by the `AppEnvironment`.

**Reply by bielikb** [reply]
The env dependency tree looks like this:

[appEnvironment](https://github.com/bielikb/scrumdinger/blob/23c35919453186cdfe1e9d2dd074654cb4fd3a1e/Scrumdinger/ScrumdingerApp.swift#L66)
👇
[scrumsViewEnvironment](https://github.com/bielikb/scrumdinger/blob/23c35919453186cdfe1e9d2dd074654cb4fd3a1e/Scrumdinger/ScrumsView.swift#L59)
👇
[detailViewEnvironment](https://github.com/bielikb/scrumdinger/blob/23c35919453186cdfe1e9d2dd074654cb4fd3a1e/Scrumdinger/DetailView.swift#L71)
👇
[meetingViewEnvironment](https://github.com/bielikb/scrumdinger/blob/23c35919453186cdfe1e9d2dd074654cb4fd3a1e/Scrumdinger/MeetingView.swift#L118)

_**where**_
`meetingViewEnvironment` instantiates the `ScrumTimer`. `ScrumTimer` is though needed only in the case the user ever gets to the `MeetingView` and shouldn't be initialised every time the app starts.

**Reply by stephencelis** [reply]
Take a look at how [ComposableCoreLocation](https://github.com/pointfreeco/composable-core-location) and [ComposableCoreMotion](https://github.com/pointfreeco/composable-core-motion) manage their life cycles. They have `.create` and `.destroy` lifecycle effects that you can fire off when someone enters and leaves a feature. Your app will still always have one of these clients in its environment, but the manager under the hood that listens for location/motion updates isn't always active.

**Reply by bielikb** [reply]
Thanks for your help. That basically answers my question.
